File: src/com/tao/chinachuclient/RuleListAdapter.java
package com.tao.chinachuclient;

import Chinachu4j.Rule;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

public class RuleListAdapter extends ArrayAdapter<Rule>{
	private LayoutInflater mInflater;

	public RuleListAdapter(Context context){
		super(context, android.R.layout.simple_list_item_1);
		mInflater = (LayoutInflater)context.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
	}

	class ViewHolder{
		TextView title, channel;
	}

	@SuppressLint("InflateParams")
	@Override
	public View getView(int position, View convertView, ViewGroup parent){
		final ViewHolder holder;
		final Rule item = getItem(position);

		if(convertView == null) {
			convertView = mInflater.inflate(R.layout.program_list_layout, null);
			TextView title = (TextView)convertView.findViewById(R.id.program_title);
			TextView channel = (TextView)convertView.findViewById(R.id.program_date);

			holder = new ViewHolder();
			holder.title = title;
			holder.channel = channel;

			convertView.setTag(holder);
		}else{
			holder = (ViewHolder)convertView.getTag();
		}

		if(item.getCategories().length > 0) {
			switch(item.getCategories()[0]){
			case "anime":
				convertView.setBackgroundResource(R.drawable.anime);
				break;
			case "information":
				convertView.setBackgroundResource(R.drawable.information);
				break;
			case "news":
				convertView.setBackgroundResource(R.drawable.news);
				break;
			case "sports":
				convertView.setBackgroundResource(R.drawable.sports);
				break;
			case "variety":
				convertView.setBackgroundResource(R.drawable.variety);
				break;
			case "drama":
				convertView.setBackgroundResource(R.drawable.drama);
				break;
			case "music":
				convertView.setBackgroundResource(R.drawable.music);
				break;
			case "cinema":
				convertView.setBackgroundResource(R.drawable.cinema);
				break;
			case "etc":
				convertView.setBackgroundResource(R.drawable.etc);
				break;
			default:
				break;
			}
		}

		String[] titles = item.getReserve_titles();
		String title = "";
		if(titles.length != 0) {
			for(String s : titles)
				title += s + ", ";
			title = title.substring(0, title.length() - 2);
		}else
			title = "any";

		String[] channels = item.getChannels();
		String channel = "";
		if(channels.length != 0) {
			for(String s : channels)
				channel += s + ", ";
			channel = channel.substring(0, channel.length() - 2);
		}else{
			channel = "any";
		}

		holder.title.setText(title);
		holder.channel.setText(channel);

		return convertView;
	}
}

File: app/src/main/java/com/tao/chinachuclient/RuleListAdapter.kt
package com.tao.chinachuclient

import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.core.content.ContextCompat
import com.tao.chinachuclient.databinding.ProgramListLayoutBinding
import sugtao4423.library.chinachu4j.Rule

class RuleListAdapter(context: Context) :
    ArrayAdapter<Rule>(context, android.R.layout.simple_list_item_1) {

    private val oldCategoryColor =
        (context.applicationContext as App).currentServer.oldCategoryColor

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding = if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            val tBinding = ProgramListLayoutBinding.inflate(inflater, parent, false)
            tBinding.root.tag = tBinding
            tBinding
        } else {
            convertView.tag as ProgramListLayoutBinding
        }

        val item = getItem(position) ?: return binding.root

        if (item.categories.isNotEmpty()) {
            binding.root.setBackgroundResource(
                if (oldCategoryColor) {
                    when (item.categories[0]) {
                        "anime" -> R.drawable.old_anime
                        "information" -> R.drawable.old_information
                        "news" -> R.drawable.old_news
                        "sports" -> R.drawable.old_sports
                        "variety" -> R.drawable.old_variety
                        "drama" -> R.drawable.old_drama
                        "music" -> R.drawable.old_music
                        "cinema" -> R.drawable.old_cinema
                        "etc" -> R.drawable.old_etc
                        else -> R.drawable.old_etc
                    }
                } else {
                    when (item.categories[0]) {
                        "anime" -> R.drawable.anime
                        "information" -> R.drawable.information
                        "news" -> R.drawable.news
                        "sports" -> R.drawable.sports
                        "variety" -> R.drawable.variety
                        "drama" -> R.drawable.drama
                        "music" -> R.drawable.music
                        "cinema" -> R.drawable.cinema
                        "etc" -> R.drawable.etc
                        else -> R.drawable.etc
                    }
                }
            )
        }

        val title = item.reserveTitles.let { if (it.isEmpty()) "any" else it.joinToString() }
        val channel = item.channels.let { if (it.isEmpty()) "any" else it.joinToString() }
        binding.programTitle.text = title
        binding.programDate.text = channel

        if (item.isDisabled) {
            binding.programTitle.let {
                it.setTextColor(Color.GRAY)
                it.paint.flags = it.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            }
            binding.programDate.let {
                it.setTextColor(Color.GRAY)
                it.paint.flags = it.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            }
        } else {
            binding.programTitle.let {
                it.setTextColor(ContextCompat.getColor(context, R.color.titleText))
                it.paint.flags = it.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
            }
            binding.programDate.let {
                it.setTextColor(ContextCompat.getColor(context, R.color.dateText))
                it.paint.flags = it.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
            }
        }

        return binding.root
    }

    override fun addAll(rules: Array<Rule>) {
        rules.map {
            add(it)
        }
    }

}
